### ISSUE TITLE
propagating_vgws not recognized as a list inside module

### ISSUE THREAD

**Issue by dmrzzz** [hashicorp/terraform#11453]
I have a module which takes a list variable and uses it to set propagating_vgws on an aws_route_table.  When I call this module and provide a list value which contains the ID of an aws_vpn_gateway, then `terraform plan` fails with "aws_route_table.rtb: propagating_vgws: should be a list".

(It's fine if I pass a hardcoded list value to the module instead, and it's also fine if I construct the aws_route_table directly in the main file using the ID of the aws_vpn_gateway)

testmodule/module.tf:
```hcl
variable vpc_id {}
variable propagating_vgws { type = "list" }

resource "aws_route_table" "rtb" {
    vpc_id = "${var.vpc_id}"
    propagating_vgws = "${var.propagating_vgws}"
}
```

main.tf:
```hcl
provider "aws" {
  region = "us-east-2"
}

variable vpc_id { default = "bogusvpc" }
variable hardcoded_vgw { default = "bogusvgw" }

resource "aws_vpn_gateway" "vgw" {
  vpc_id = "${var.vpc_id}"
}

# OK
resource "aws_route_table" "campus-a" {
  vpc_id = "${var.vpc_id}"
  propagating_vgws = ["${aws_vpn_gateway.vgw.id}"]
}

# OK
module "module1" {
  source = "testmodule"
  vpc_id = "${var.vpc_id}"
  propagating_vgws = []
}

# OK
module "module2" {
  source = "testmodule"
  vpc_id = "${var.vpc_id}"
  propagating_vgws = ["${var.hardcoded_vgw}"]
}

# terraform plan fails with error
module "module3" {
  source = "testmodule"
  vpc_id = "${var.vpc_id}"
  propagating_vgws = ["${aws_vpn_gateway.vgw.id}"]
}
```

```
dmrz@golbez:~/tmp/tftest$ terraform -v
Terraform v0.8.5

dmrz@golbez:~/tmp/tftest$ tree
.
├── main.tf
└── testmodule
    └── module.tf

1 directory, 2 files
dmrz@golbez:~/tmp/tftest$ terraform get
Get: file:///home/dmrz/tmp/tftest/testmodule
Get: file:///home/dmrz/tmp/tftest/testmodule
Get: file:///home/dmrz/tmp/tftest/testmodule
dmrz@golbez:~/tmp/tftest$ terraform plan
There are warnings and/or errors related to your configuration. Please
fix these before continuing.

Errors:

  * aws_route_table.rtb: propagating_vgws: should be a list
```

**Comment by dmrzzz** [comment]
Same problem with alarm_actions and friends in aws_cloudwatch_metric_alarm

**Comment by rsalmond** [comment]
I just helped a colleague with this, I'm not sure why this works but this is the way to pass lists into a module. You need to wrap the variable inside square brackets in the module that uses it.

```hcl
variable vpc_id {}
variable propagating_vgws { type = "list" }

resource "aws_route_table" "rtb" {
    vpc_id = "${var.vpc_id}"
    propagating_vgws = ["${var.propagating_vgws}"]
}
```

**Comment by teamterraform** [comment]
Hi all! Sorry for the long silence here.

This was a common problem in Terraform 0.11 because of a limitation of how it represented values: any list that contained a value that would not be known until apply would become a wholly-unknown list, and (due to a representation ambiguity) an unknown value could not be recognized as a list by the type checking rules in the provider SDK.

Terraform 0.12 has fixed this by changing how "unknown" is represented. It's now possible to have a list containing a mixture of known and unknown elements, and additionally Terraform can now track the type of a value even if the actual value is unknown. That is, it can represent "unknown list of strings" specifically, whereas before it could only represent "unknown value of unknown type".

As @rsalmond noted, it was possible to trick Terraform 0.11 into accepting an unknown list by placing a redundant set of list brackets around it. That workaround was an unintended consequence of a backward-compatibility feature added in Terraform 0.7, so it was a lucky coincidence that it happened to avoid the type checking problem, rather than an intended feature.

As a consequence, this issue is no longer present in Terraform 0.12. The aforementioned Terraform 0.7 compatibility feature is no longer present in Terraform 0.12, so if you _did_ add redundant list brackets to work around this problem then you'll need to remove them when upgrading to Terraform 0.12 or it will treat the expression as a list of lists. The `terraform 0.12upgrade` tool is able to recognize and automatically fix that in some cases, but it doesn't always have enough information to know that rewrite is needed. If there are cases the upgrade tool doesn't catch, a subsequent Terraform run will detect it and raise an error similar to the following:

```
Error: Incorrect attribute value type

  on redundant-list-brackets.tf line 9, in resource "aws_security_group" "foo":
   9:     cidr_blocks = ["${var.cidr_blocks}"]

Inappropriate value for attribute "cidr_blocks": element 0: string required.
````

If you see that, the usual fix is to remove those redundant list brackets, and let Terraform's own type checker detect that the resulting expression is itself a list.

Because this is fixed in Terraform 0.12, we're going to close this out now. Thanks for reporting this, and sorry again for the long delay in responding to it.
